proposals/threads/Globals.md

```diff
@@ -2,6 +2,316 @@
+  function entrypoints
+* Additional thread-local values must also be spilled/loaded in the same way to
+  have the same benefit
+* The SP is still in linear memory, so it can easily be trashed by another
+  agent
+
+### Solution 3: Modify Function Signature to Pass SP as Parameter
+
+Rather than storing spilling the SP to linear memory, the SP value can be
+passed as a parameter. Because we can't tell whether an imported function will
+use the SP, we must modify all imported and exported functions.
+
+The SP will ultimately be saved in a mutable global, but will be loaded from
+the parameter at function entrypoints:
+
+```
+(module $m1
+  (memory (export "memory") 1)
+  (global $sp (mut i32) (i32.const 0))
+
+  (func $exported (export "exported") (param $sp i32)
+    ;; Load SP from param
+    (set_global $sp (get_local $sp))
+
+    ;; SP = SP + 64
+    (set_global $sp (i32.add (get_global $sp) (i32.const 64)))
+    ...
+  )
+  ...
+)
+
+(module $m2
+  (import "env" "memory" (memory 1))
+  (import "env" "exported" (func $exported (param $sp i32)))
+
+  (global $sp (mut i32) (i32.const 0))
+
+  (func $internal
+    ;; SP doesn't need to be loaded because this function is internal
+
+    ;; SP = SP + 4
+    (set_global $sp (i32.add (get_global $sp) (i32.const 4)))
+
+    (call $exported (get_global $sp))
+  )
+)
+```
+
+The modules can then be instantiated as follows:
+
+```
+WebAssembly.instantiate(m1Bytes, {}).then(
+    ({_, instance}) => {
+        let imports = {env: {
+          memory: instance.exports.memory,
+          exported: instance.exports.exported,
+        }};
+        WebAssembly.instantiate(m2Bytes, imports).then(...);
+    });
+```
+
+But now the JavaScript code must keep a global SP that can be passed to
+exported functions. In addition, any functions that call back to JavaScript
+must have this SP updated:
+
+```
+let sp = 0x200;
+
+function importedFunction(newSp) {
+  sp = newSp;
+  ...
+}
+
+m1.exports.exported(sp);
+```
+
+The drawback for this solution is that all exported functions must have an
+additional parameter for each thread-local value. This solution could be
+extended for other thread-local values, but will very quickly become unwieldy.
+
+### Proposed Solution: Import and Export Mutable Globals
+
+In the MVP, mutable globals cannot be imported or exported. If we loosen this
+restriction, we can provide a much nicer solution for thread-local values:
+
+```
+(module $m1
+  (import "env" "sp" (global $sp (mut i32)))
+  (memory (export "memory") 1)
+
+  (func
+    ;; SP = SP + 64
+    (set_global $sp (i32.add (get_global $sp) (i32.const 64)))
+    ...
+  )
+  ...
+)
+
+(module $m2
+  (import "env" "memory" (memory 1))
+  (import "env" "sp" (global $sp (mut i32)))
+  ...
+
+  (func
+    ;; SP = SP + 4
+    (set_global $sp (i32.add (get_global $sp (i32.const 4))))
+  )
+)
+```
+
+With the modules instantiated as follows:
+
+```
+let agentSp = 0x100;
```

Comment:
This example is interesting because it doesn't show any usage of `WebAssembly.Global`. Is that a mistake? Or do we need another example to show the value-add of that?